File: week2/python/hueHistogram.py
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
# 
# This code is made available to the students of 
# the online course titled "Computer Vision for Faces" 
# by Satya Mallick for personal non-commercial use. 
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC. 
#
# For licensing and other inquiries, please email 
# spmallick@bigvisionllc.com 
# 

import cv2,argparse
import numpy as np

ap = argparse.ArgumentParser()
ap.add_argument("-f", "--filename")
args = vars(ap.parse_args())

filename = "../data/images/capsicum.jpg"
if args['filename']:
  filename = args['filename']

img = cv2.imread(filename)

# Convert to HSV color space
hsvImage = cv2.cvtColor(img,cv2.COLOR_BGR2HSV)

# Convert to float32
hsvImage = np.float32(hsvImage)

# Split the channels
H, S, V = cv2.split(hsvImage)

# Specifying the width and height of the figure used to plot the histogram
plot_width = 540
plot_height = 400
actualRange = 180
rangeRatio = int(plot_width/actualRange)
bufferHeight = 30


# Create an empty image to plot the histogram
histImage = 255*np.ones((plot_height + bufferHeight, plot_width, 3))
xAxisValues = np.arange(plot_width)

# Find the histogram
histogram = cv2.calcHist([H*rangeRatio],[0],None,[plot_width],[0,plot_width])

# Normalize the histogram so that y-values do not overshoot the height of the figure 
cv2.normalize(histogram, histogram, 0, plot_height, cv2.NORM_MINMAX, -1)

# Normalize the histogram to be plotted as an image
histogram = np.max(histogram) - histogram

# Create points using x-axis and y-axis arrays
points = np.column_stack((xAxisValues,histogram.T[0]))

# Draw the histogram as lines on the empty image
cv2.polylines(histImage, np.int32([points]), False, (0,0,255), 2, cv2.LINE_AA)

# Also draw the x-axis values at an interval given by xInterval
# parameters used in putText
font = cv2.FONT_HERSHEY_SIMPLEX
fontScale = 0.4
fontWidth = 1
xInterval = 20

# Special case for putting the origin of x axis
cv2.putText(histImage, "0", (0, int(histImage.shape[0] - (bufferHeight/2)) ), font, fontScale, (0, 0, 0), fontWidth)

for i in range(0,plot_width,rangeRatio*xInterval):
  
  # Specify the position ( xval, yval )for putting the point on x axis 
  xval = i - 7  
  yval = int(histImage.shape[0] - (bufferHeight/2))
  cv2.putText(histImage, str( i/rangeRatio ), (xval, yval ), font, fontScale, (0, 0, 0), fontWidth)

cv2.namedWindow("Original Image", cv2.WINDOW_AUTOSIZE)
cv2.namedWindow("Histogram of Hue channel", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Original Image",img)
cv2.imshow("Histogram of Hue channel",histImage)
cv2.waitKey(0)
cv2.destroyAllWindows()
cv2.imwrite('results/hueHistogram.png',histImage)

File: week2/cpp/hueHistogram.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "Computer Vision for Faces" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/

#include <iostream>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
 
using namespace std;
using namespace cv;

int main(int argc, char ** argv)
{

  // Read the image
  string filename = "../data/images/capsicum.jpg";
  if (argc == 2)
  {   
    filename = argv[1];
  }
  
  Mat img = imread(filename);

  // Convert to HSV color space
  Mat hsvImage;
  cvtColor(img, hsvImage, COLOR_BGR2HSV);

  // Split the channels
  vector<Mat> channels(3);
  split(hsvImage, channels);
  
  imshow( "Image", img );

  // Initialize parameters
  int histSize = 180;    // bin size
  float range[] = { 0, 179 };
  const float *ranges[] = { range };

  // Calculate histogram
  MatND hist;
  calcHist( &channels[0], 1, 0, Mat(), hist, 1, &histSize, ranges, true, false );

  // Parameters for the plot
  int hist_w = histSize*3; int hist_h = 400;
  int bin_w = cvRound( (double) hist_w/histSize );
  
  // construct the histogram as an image
  Mat histImage( hist_h, hist_w, CV_8UC3, Scalar( 255,255,255) );
  normalize(hist, hist, 0, histImage.rows, NORM_MINMAX, -1, Mat() );
  
  // draw the x-axis
  line(histImage, Point(0, hist_h - 30), Point(hist_w, hist_h - 30), Scalar(0, 0, 0), 2, 8, 0);
  
  // Special case for specifying the origin of x-axis
  line(histImage, Point(0, hist_h - 35), Point(0, hist_h - 25), Scalar(0, 0, 0), 2, 8, 0);
  putText(histImage, "0", Point(0, hist_h-5), cv::FONT_HERSHEY_COMPLEX, .5, (0,0,0), 1, cv::LINE_AA);
  // Draw the histogram
  for( int i = 1; i < histSize; i++ )
  {
    line( histImage, Point( bin_w*(i-1), hist_h - 30 - cvRound(hist.at<float>(i-1)) ) ,
                     Point( bin_w*(i), hist_h - 30 - cvRound(hist.at<float>(i)) ),
                     Scalar( 0, 0, 255), 2, 8, 0  );
    
    // show the x axis values
    if (i % 20 == 0)
    {
      char buffer[5];
      sprintf(buffer,"%d",i);
      line(histImage, Point(i*bin_w, hist_h - 35), Point(i*bin_w, hist_h - 25), Scalar(0, 0, 0), 2, 8, 0);
      putText(histImage, buffer, Point(i*bin_w, hist_h-5), cv::FONT_HERSHEY_COMPLEX, .5, (0,0,0), 1, cv::LINE_AA);
    }
  }
  namedWindow("Original Image",CV_WINDOW_AUTOSIZE);
  namedWindow("Histogram of Hue channel",CV_WINDOW_AUTOSIZE);

  imshow("Original Image", img);
  imshow( "Histogram of Hue channel", histImage );
  imwrite( "results/hueHistogram.jpg", histImage );

  waitKey(0);   
  destroyAllWindows(); 
  return 0;
}
